src/main/java/org/mockbukkit/mockbukkit/attribute/AttributeInstanceMock.java

```diff
@@ -20,111 +23,156 @@
-	{
-		return attribute;
-	}
-
-	@Override
-	public double getBaseValue()
-	{
-		return value;
-	}
-
-	@Override
-	public void setBaseValue(double value)
-	{
-		this.value = value;
+		this.defaultValue = defaultValue;
+		this.baseValue = defaultValue;
 	}
 
 	@Override
 	public @NotNull Collection<AttributeModifier> getModifiers()
 	{
-		return modifiers;
+		return Collections.unmodifiableCollection(modifiers);
 	}
 
 	@Override
-	public @Nullable AttributeModifier getModifier(@NotNull Key key)
+	public @Nullable AttributeModifier getModifier(@NotNull net.kyori.adventure.key.Key key)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		Preconditions.checkNotNull(key, "Key cannot be null");
+		return modifiersByKey.get(key);
 	}
 
 	@Override
-	public void removeModifier(@NotNull Key key)
+	public void removeModifier(@NotNull net.kyori.adventure.key.Key key)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		Preconditions.checkNotNull(key, "Key cannot be null");
+		AttributeModifier modifier = modifiersByKey.remove(key);
+		if (modifier != null)
+		{
+			modifiers.remove(modifier);
+			modifiersByUuid.remove(modifier.getUniqueId());
+		}
 	}
 
 	@Override
 	@Deprecated(forRemoval = true, since = "1.21")
 	public @Nullable AttributeModifier getModifier(@NotNull UUID uuid)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		Preconditions.checkNotNull(uuid, "UUID cannot be null");
+		return modifiersByUuid.get(uuid);
 	}
 
 	@Override
 	@Deprecated(forRemoval = true, since = "1.21")
 	public void removeModifier(@NotNull UUID uuid)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		Preconditions.checkNotNull(uuid, "UUID cannot be null");
+		AttributeModifier modifier = modifiersByUuid.remove(uuid);
+		if (modifier != null)
+		{
+			modifiers.remove(modifier);
+			modifiersByKey.remove(modifier.getKey());
+		}
 	}
 
 	@Override
 	public void addModifier(@NotNull AttributeModifier modifier)
 	{
-		Preconditions.checkNotNull(modifier, "Modifier cannot be null");
+		Preconditions.checkNotNull(modifier, "Modifier shouldn't be null");
+
+		// Remove existing modifier with same key or UUID if present
+		removeModifier(modifier.getKey());
+		removeModifier(modifier.getUniqueId());
+
 		modifiers.add(modifier);
+		modifiersByKey.put(modifier.getKey(), modifier);
+		modifiersByUuid.put(modifier.getUniqueId(), modifier);
 	}
 
 	@Override
 	public void addTransientModifier(@NotNull AttributeModifier modifier)
 	{
-		// TODO Auto-generated method stub
-		throw new UnimplementedOperationException();
+		Preconditions.checkNotNull(modifier, "Modifier cannot be null");
+
+		// Remove from transient if already exists
+		transientModifiers.removeIf(m ->
+				m.getKey().equals(modifier.getKey()) ||
+						m.getUniqueId().equals(modifier.getUniqueId()));
+
+		transientModifiers.add(modifier);
```

Comment:
From what I can understand, transient modifiers should be stored in the same collection/map. That's how the NMS implementation is.

I actually think you don't even need to differentiate between these two types, as there is no world storage solution for MockBukkit; This does not matter, as nothing is persistent

Reply:
removed the transient implementation